# Robin Nguyen

# I. Insert data 
## 1. Log Daily Return

In [10]:
# Import packages
import os
import yfinance as yf
import numpy as np
import pandas as pd
import warnings

In [17]:
# Define a function to generate Yahoo Finance Data
def dret(ticker, start_date="2018-01-01", end_date="2024-12-31", output_directory=output_dir):

    # Create the output directory if it doesn't exist
    if not os.path.exists(output_directory):
        os.makedirs(output_directory)

    # Download price data
    pricedata = yf.download(ticker, start=start_date, end=end_date, progress=False)

    if pricedata.empty:
        return None

    # Suppress RuntimeWarning: invalid value encountered in log
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", category=RuntimeWarning)
        
        # Calculate daily log returns using the 'Close' column
        pricedata[('Daily Return', ticker)] = np.log(pricedata['Close'] / pricedata['Close'].shift(1))

    # Drop NaN values resulting from the shift
    pricedata = pricedata.dropna()

    # Define the output file path
    output_file = f"{output_directory}/{ticker}.csv"

    # Check if the file exists and delete it before writing
    if os.path.exists(output_file):
        os.remove(output_file)  # Delete the file to avoid permission issues

    # Save the data to the file
    pricedata.to_csv(output_file, index=True)

    return pricedata

In [18]:
# Define the directory path for saving data
output_dir_tech = "./data/quantitative/tech"

# List of stock tickers
tech_tickers = [
    "AAPL", "MSFT", "GOOGL", "GOOG", "AMZN", "META", "NVDA", "TSM", "ADBE", "INTC",
    "CSCO", "ORCL", "IBM", "CRM", "QCOM", "AVGO", "TXN", "AMD", "AMAT", "MU",
    "NET", "NOW", "SNOW", "DOCU", "SHOP", "UBER", "LYFT", "SNAP", "HRB", "DDOG"
]

# Download data for each ticker
for ticker in tech_tickers:
    data_tech = dret(ticker, start_date="2018-01-01", end_date="2024-12-31", output_directory=output_dir_tech)

YF.download() has changed argument auto_adjust default to True


In [25]:
# Define the directory path for saving data
output_dir_finbank = "./data/quantitative/fin_bank"

# List of financial and banking tickers
finbank_tickers = [
    "JPM", "BAC", "WFC", "C", "GS", "MS", "USB", "PNC", "TFC", "COF",
    "TD", "SCHW", "BK", "STT", "AXP", "HSBC", "CFG", "FITB", "MTB", "HBAN",
    "ALLY", "KEY", "RY", "SAN", "NTRS", "RF", "SYF", "NBHC", "ZION", "FHN"
]

# Download data for each ticker
for ticker in finbank_tickers:
    data_finbank = dret(ticker, start_date="2018-01-01", end_date="2024-12-31", output_directory=output_dir_finbank)

In [26]:
# Define the directory path for saving data
output_dir_ev = "./data/quantitative/ev"

# List of financial and banking tickers
ev_tickers = [
    "TSLA", "BYDDY", "LI", "NIO", "RIVN", "LCID", "XPEV", "NKLA", "PSNY", "GM", "F",
    "VWAGY", "BAMXF", "HYMTF", "KIMTF", "POAHY", "MBGYY", "STLA", "GELYF", "GWLLY",
    "SAIC", "HYLN", "GNZUF", "TATAMOTORS.NS", "MAHMF", "RNLSY", "NSANY", "MMTOF"
]

# Download data for each ticker
for ticker in ev_tickers:
    data_ev = dret(ticker, start_date="2018-01-01", end_date="2024-12-31", output_directory=output_dir_ev)

In [27]:
# Define the directory path for saving data
output_dir_market = "./data/quantitative"

market_data = dret("^GSPC", start_date="2018-01-01", end_date="2024-12-31", output_directory=output_dir_market)

In [24]:
# View data
market_data

Price             Close         High          Low         Open      Volume  \
Ticker            ^GSPC        ^GSPC        ^GSPC        ^GSPC       ^GSPC   
Date                                                                         
2018-01-03  2713.060059  2714.370117  2697.770020  2697.850098  3544030000   
2018-01-04  2723.989990  2729.290039  2719.070068  2719.310059  3697340000   
2018-01-05  2743.149902  2743.449951  2727.919922  2731.330078  3239280000   
2018-01-08  2747.709961  2748.510010  2737.600098  2742.669922  3246160000   
2018-01-09  2751.290039  2759.139893  2747.860107  2751.149902  3467460000   
...                 ...          ...          ...          ...         ...   
2024-12-23  5974.069824  5978.250000  5902.569824  5940.250000  3593280000   
2024-12-24  6040.040039  6040.100098  5981.439941  5984.629883  1757720000   
2024-12-26  6037.589844  6049.750000  6007.370117  6024.970215  2904530000   
2024-12-27  5970.839844  6006.169922  5932.950195  6006.169922  3159610000   
2024-12-30  5906.939941  5940.790039  5869.160156  5920.669922  3433250000   

Price      Daily Return  
Ticker            ^GSPC  
Date                     
2018-01-03     0.006378  
2018-01-04     0.004021  
2018-01-05     0.007009  
2018-01-08     0.001661  
2018-01-09     0.001302  
...                 ...  
2024-12-23     0.007261  
2024-12-24     0.010982  
2024-12-26    -0.000406  
2024-12-27    -0.011117  
2024-12-30    -0.010760  

[1759 rows x 6 columns]

## 2. FAMA-FERNCH 3-FACTOR MODEL

In [5]:
fama_french_df = pd.read_csv(r"C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\F-F_Research_Data_Factors_daily.CSV",header=0, skiprows=3, sep=",")

In [6]:
# view dataframe
fama_french_df

Unnamed: 0  Mkt-RF   SMB   HML  \
0                                               19260701    0.10 -0.25 -0.27   
1                                               19260702    0.45 -0.33 -0.06   
2                                               19260706    0.17  0.30 -0.39   
3                                               19260707    0.09 -0.58  0.02   
4                                               19260708    0.21 -0.38  0.19   
...                                                  ...     ...   ...   ...   
25897                                           20241226    0.02  1.04 -0.19   
25898                                           20241227   -1.17 -0.66  0.56   
25899                                           20241230   -1.09  0.12  0.74   
25900                                           20241231   -0.46  0.00  0.71   
25901  Copyright 2024 Eugene F. Fama and Kenneth R. F...     NaN   NaN   NaN   

          RF  
0      0.009  
1      0.009  
2      0.009  
3      0.009  
4      0.009  
...      ...  
25897  0.017  
25898  0.017  
25899  0.017  
25900  0.017  
25901    NaN  

[25902 rows x 5 columns]

In [9]:

# Rename the 'Unnamed: 0' column to 'Date'
fama_french_df = fama_french_df.rename(columns={'Unnamed: 0': 'Date'})

# Remove the copyright row at the end of the table
fama_french_df = fama_french_df.iloc[:-1, :]

# Convert the 'Date' column to the appropriate datetime format
fama_french_df['Date'] = pd.to_datetime(fama_french_df['Date'], format='%Y%m%d')

# Convert percentage columns to decimal numbers
numeric_columns = ['Mkt-RF', 'SMB', 'HML', 'RF']
fama_french_df[numeric_columns] = fama_french_df[numeric_columns].astype(float) / 100

# Filter the data for the specified date range
start_date = pd.Timestamp('2018-01-01')
end_date = pd.Timestamp('2024-12-31')
fama_french_df = fama_french_df.loc[(fama_french_df['Date'] >= start_date) & (fama_french_df['Date'] <= end_date)]

# Display the first few rows of the filtered DataFrame
print(fama_french_df.head())

# Save the cleaned DataFrame to a new CSV file
output_file = r"C:\\Users\\haloi\\Downloads\\data_201_proj\\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\\data\\quantitative\\fama_french\\Cleaned_Fama_French_Data.csv"
fama_french_df.to_csv(output_file, index=False)

print(f"Cleaned data has been saved to: {output_file}")

            Date    Mkt-RF       SMB       HML            RF
24140 2018-01-02  0.000085  0.000035 -0.000022  6.000000e-07
24141 2018-01-03  0.000059 -0.000039 -0.000018  6.000000e-07
24142 2018-01-04  0.000042 -0.000026  0.000024  6.000000e-07
24143 2018-01-05  0.000066 -0.000036 -0.000026  6.000000e-07
24144 2018-01-08  0.000019 -0.000015  0.000004  6.000000e-07
Cleaned data has been saved to: C:\\Users\\haloi\\Downloads\\data_201_proj\\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\\data\\quantitative\\fama_french\\Cleaned_Fama_French_Data.csv


In [8]:
# summary satistic
fama_french_df.iloc[:, 1:5].describe()

Mkt-RF          SMB          HML           RF
count  1761.000000  1761.000000  1761.000000  1761.000000
mean      0.000488    -0.000078    -0.000121     0.000089
std       0.012738     0.007145     0.010180     0.000078
min      -0.120000    -0.035600    -0.049700     0.000000
25%      -0.004800    -0.004400    -0.005500     0.000000
50%       0.000800    -0.000300    -0.000500     0.000070
75%       0.007100     0.004100     0.005200     0.000170
max       0.093400     0.054700     0.067300     0.000220

In [191]:
# Check for missing values
missing_values = fama_french_df.isna().sum()
print("Missing Values:")
print(missing_values[missing_values > 0])

Missing Values:
Series([], dtype: int64)


In [12]:
import pandas as pd
import os

# Define paths
stock_data_dir = r"C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\ev"
fama_french_file_path = r"C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\Cleaned_Fama_French_Data.csv"
output_dir = r"C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev"

# Load the Fama-French data
fama_french_df = pd.read_csv(fama_french_file_path)
fama_french_df['Date'] = pd.to_datetime(fama_french_df['Date'])  # Ensure 'Date' is in datetime format

# Ensure the output directory exists
os.makedirs(output_dir, exist_ok=True)

# Process each stock data file in the stock data directory
for file_name in os.listdir(stock_data_dir):
    if file_name.endswith('.csv'):  # Ensure it's a CSV file
        stock_file_path = os.path.join(stock_data_dir, file_name)

        try:
            # Load the stock data with the first three rows as header levels
            stock_df = pd.read_csv(stock_file_path, header=[0, 1, 2])  # Include third row for column names
            
            # Flatten the multi-level columns
            stock_df.columns = stock_df.columns.map(lambda x: x[2])  # Use the third level of the header for column names
            
            # Rename the columns
            new_column_names = ['Date', 'Close', 'High', 'Low', 'Open', 'Volume', 'Daily Return']
            if len(stock_df.columns) == len(new_column_names):
                stock_df.columns = new_column_names
            else:
                print(f"Column count mismatch in {file_name}: Skipping file.")
                continue  # Skip this file if column count doesn't match
            
            # Ensure 'Date' is in datetime format
            stock_df['Date'] = pd.to_datetime(stock_df['Date'])
            
            # Merge stock data with Fama-French factors
            merged_df = pd.merge(stock_df, fama_french_df, on='Date', how='inner')
            
            # Save the merged DataFrame to the output directory
            output_file_path = os.path.join(output_dir, f"merged_{file_name}")
            merged_df.to_csv(output_file_path, index=False)

            print(f"Merged data for {file_name} saved to: {output_file_path}")
        
        except Exception as e:
            print(f"Error processing {file_name}: {e}")

Merged data for AAPL.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_AAPL.csv
Merged data for ADBE.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_ADBE.csv
Merged data for AMAT.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_AMAT.csv
Merged data for AMD.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_AMD.csv
Merged data for AMZN.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_A

Merged data for QCOM.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_QCOM.csv
Merged data for RIVN.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_RIVN.csv
Merged data for RNLSY.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_RNLSY.csv
Merged data for SAIC.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merged_SAIC.csv
Merged data for SHOP.csv saved to: C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev\merg

In [13]:
import pandas as pd
import os
import statsmodels.api as sm
from sklearn.metrics import mean_absolute_error, mean_squared_error
import numpy as np

# Define the path to the directory with merged files
merged_data_dir = r"C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev"

# Loop through all merged CSV files in the directory
for file_name in os.listdir(merged_data_dir):
    if file_name.startswith("merged_") and file_name.endswith(".csv"):  # Process only merged files
        merged_file_path = os.path.join(merged_data_dir, file_name)

        try:
            # Load the merged CSV file
            df = pd.read_csv(merged_file_path)
            print(f"Processing file: {file_name}")

            # Ensure 'Date' is in datetime format
            df['Date'] = pd.to_datetime(df['Date'])

            # Calculate y (dependent variable): Daily Return - RF
            df['y'] = df['Daily Return'] - df['RF']

            # Define x (independent variables): Mkt-RF, SMB, HML
            x = df[['Mkt-RF', 'SMB', 'HML']]
            y = df['y']

            # Add a constant term to the independent variables for the intercept
            x = sm.add_constant(x)

            # Perform the regression
            model = sm.OLS(y, x).fit()

            # Predicted values
            y_pred = model.predict(x)

            # Calculate verification metrics
            mae = mean_absolute_error(y, y_pred)
            mse = mean_squared_error(y, y_pred)
            rmse = np.sqrt(mse)

            # Save the regression summary and metrics to a text file
            output_summary_path = os.path.join(merged_data_dir, f"fama_french_model_summary_{file_name.replace('merged_', '').replace('.csv', '')}.txt")
            with open(output_summary_path, "w") as f:
                f.write(model.summary().as_text())
                f.write("\n\nVerification Metrics:\n")
                f.write(f"Mean Absolute Error (MAE): {mae:.6f}\n")
                f.write(f"Mean Squared Error (MSE): {mse:.6f}\n")
                f.write(f"Root Mean Squared Error (RMSE): {rmse:.6f}\n")
        
        except Exception as e:
            print(f"Error processing {file_name}: {e}")

Processing file: merged_AAPL.csv
Processing file: merged_ADBE.csv
Processing file: merged_AMAT.csv
Processing file: merged_AMD.csv
Processing file: merged_AMZN.csv
Processing file: merged_AVGO.csv
Processing file: merged_BAMXF.csv
Processing file: merged_BYDDY.csv
Processing file: merged_CRM.csv
Processing file: merged_CSCO.csv
Processing file: merged_DDOG.csv
Processing file: merged_DOCU.csv
Processing file: merged_F.csv
Processing file: merged_GELYF.csv
Processing file: merged_GM.csv
Processing file: merged_GNZUF.csv
Processing file: merged_GOOG.csv
Processing file: merged_GOOGL.csv
Processing file: merged_GWLLY.csv
Processing file: merged_HRB.csv
Processing file: merged_HYLN.csv
Processing file: merged_HYMTF.csv
Processing file: merged_IBM.csv
Processing file: merged_INTC.csv
Processing file: merged_KIMTF.csv
Processing file: merged_LCID.csv
Processing file: merged_LI.csv
Processing file: merged_LYFT.csv
Processing file: merged_MAHMF.csv
Processing file: merged_MBGYY.csv
Processing 

In [30]:
import os

# Define the path to the directory containing the .txt files
results_dir = r"C:\Users\haloi\Downloads\data_201_proj\Forecasting-Stock-Perfomance-using-Sentiment-Analysis-ARIMAX-RNN-and-LSTM-main\data\quantitative\fama_french\ev"

# Stocks to extract metrics for
specific_stocks = ['TSLA', 'AAPL', 'AMZN', 'ADBE']

# Loop through the specific stocks
for stock in specific_stocks:
    # Construct the file name for each stock
    file_path = os.path.join(results_dir, f"fama_french_model_summary_{stock}.txt")

    # Check if the file exists
    if os.path.exists(file_path):
        print(f"\n--- Metrics for {stock} ---")
        try:
            # Open and read the file
            with open(file_path, 'r') as file:
                lines = file.readlines()

            # Search for the metrics in the file
            for line in lines:
                if "Mean Absolute Error (MAE):" in line:
                    print(line.strip())
                elif "Mean Squared Error (MSE):" in line:
                    print(line.strip())
                elif "Root Mean Squared Error (RMSE):" in line:
                    print(line.strip())
        except Exception as e:
            print(f"Error reading file for {stock}: {e}")
    else:
        print(f"File not found for {stock}: {file_path}")


--- Metrics for TSLA ---
Mean Absolute Error (MAE): 0.023200
Mean Squared Error (MSE): 0.001112
Root Mean Squared Error (RMSE): 0.033352

--- Metrics for AAPL ---
Mean Absolute Error (MAE): 0.008069
Mean Squared Error (MSE): 0.000129
Root Mean Squared Error (RMSE): 0.011343

--- Metrics for AMZN ---
Mean Absolute Error (MAE): 0.009660
Mean Squared Error (MSE): 0.000201
Root Mean Squared Error (RMSE): 0.014161

--- Metrics for ADBE ---
Mean Absolute Error (MAE): 0.009241
Mean Squared Error (MSE): 0.000218
Root Mean Squared Error (RMSE): 0.014754


## - TSLA: The MAE of 0.0232 indicates that, on average, the model's predictions deviate from the true values by approximately 2.32%. The MSE and RMSE highlight that larger errors are present, as the RMSE (0.0334) is significantly higher than the MAE. This suggests that the model struggles with capturing extreme deviations in TSLA's stock performance.

## - AAPL:  The MAE of 0.0081 shows that the model's predictions are highly accurate, with an average error of less than 1%. The MSE (0.000129) and RMSE (0.0113) confirm strong performance, as the small values indicate the model handles both average and extreme errors well. Overall, the model performs exceptionally well for AAPL.

## - ADBE: The MAE of 0.0097 indicates that the model's predictions for AMZN deviate by less than 1% on average. The MSE (0.000201) and RMSE (0.0142) suggest that while the model has minor errors, it performs slightly worse than it does for AAPL. However, the overall accuracy is still satisfactory.

## - AMZN : The MAE of 0.0092 reflects that the model's average prediction error for ADBE is just below 1%. The MSE (0.000218) and RMSE (0.0148) suggest that the model performs slightly less effectively compared to AAPL and AMZN. While ADBE's metrics show reasonable accuracy, the slightly higher RMSE indicates occasional larger prediction errors.


# II. CAPM MODEL

In [146]:
# Import packages
import os
import pandas as pd
import statsmodels.api as sm

In [14]:
# Define CAPM function
def perform_capm(stock_tickers, market_data, output_dir, start_date="2018-01-01", end_date="2024-12-31"):
    stock_data = {}
    capm_results = {}

    # Download stock data for each ticker
    for ticker in stock_tickers:
        stock_data[ticker] = dret(ticker, start_date=start_date, end_date=end_date, output_directory=output_dir)

    # Ensure market data is available
    if market_data is not None:
        # Independent variable: market daily returns
        market_returns = market_data['Daily Return']

        # Perform CAPM regression for each stock
        for ticker, data in stock_data.items():
            if data is not None:
                try:
                    # Dependent variable: stock daily returns
                    stock_returns = data['Daily Return']

                    # Align data to avoid NaN issues
                    combined_data = pd.concat([stock_returns, market_returns], axis=1).dropna()
                    stock_returns_aligned = combined_data.iloc[:, 0]
                    market_returns_aligned = combined_data.iloc[:, 1]

                    # Add a constant to the independent variable
                    x = sm.add_constant(market_returns_aligned)

                    # Perform linear regression
                    model = sm.OLS(stock_returns_aligned, x).fit()

                    # Store results in the dictionary
                    capm_results[ticker] = model

                    # Print the summary for the ticker
                    print(f"CAPM Results for {ticker}")
                    print(model.summary())
                    print("\n" + "-" * 80 + "\n")

                except Exception as e:
                    print(f"Error processing CAPM for {ticker}: {e}")

    return capm_results

In [15]:
# Perform CAPM regression for the tickers
capm_results_tech = perform_capm(tech_tickers, market_data, output_dir_tech)
capm_results_ev = perform_capm(tech_tickers, market_data, output_dir_ev)
capm_results_finbank = perform_capm(tech_tickers, market_data, output_dir_finbank)

NameError: name 'tech_tickers' is not defined

In [149]:
# Define output direction
output_dir = "./data/quantitative/CAPM_results"

In [150]:
# Export CAPM regression summaries to a text file
output_summary_tech = os.path.join(output_dir, "CAPM_Regression_Summaries_Tech.txt")

with open(output_summary_tech, "w") as file:
    for ticker, model in capm_results_tech.items():
        file.write(f"CAPM Regression Summary for {ticker}:\n")
        file.write(model.summary().as_text())  # Write the summary as text
        file.write("\n" + "-" * 80 + "\n")  # Add a separator for readability

print(f"CAPM regression summaries have been exported to {output_summary_tech}")

CAPM regression summaries have been exported to ./data/quantitative/CAPM_results\CAPM_Regression_Summaries_Tech.txt


In [151]:
# Export CAPM regression summaries to a text file
output_summary_ev = os.path.join(output_dir, "CAPM_Regression_Summaries_Ev.txt")

with open(output_summary_ev, "w") as file:
    for ticker, model in capm_results_tech.items():
        file.write(f"CAPM Regression Summary for {ticker}:\n")
        file.write(model.summary().as_text())  # Write the summary as text
        file.write("\n" + "-" * 80 + "\n")  # Add a separator for readability

print(f"CAPM regression summaries have been exported to {output_summary_ev}")

CAPM regression summaries have been exported to ./data/quantitative/CAPM_results\CAPM_Regression_Summaries_Ev.txt


In [152]:
# Export CAPM regression summaries to a text file
output_summary_finbank = os.path.join(output_dir, "CAPM_Regression_Summaries_Finbank.txt")

with open(output_summary_finbank, "w") as file:
    for ticker, model in capm_results_tech.items():
        file.write(f"CAPM Regression Summary for {ticker}:\n")
        file.write(model.summary().as_text())  # Write the summary as text
        file.write("\n" + "-" * 80 + "\n")  # Add a separator for readability

print(f"CAPM regression summaries have been exported to {output_summary_finbank}")

CAPM regression summaries have been exported to ./data/quantitative/CAPM_results\CAPM_Regression_Summaries_Finbank.txt
